Project2015To2017/Analysis/Analyzer.cs

```diff
@@ -0,0 +1,93 @@
+using Project2015To2017.Analysis.Diagnostics;
+using Project2015To2017.Definition;
+using Project2015To2017.Reading;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Project2015To2017.Analysis
+{
+	public class Analyzer
+	{
+		public readonly IList<DiagnosticBase> Diagnostics = new List<DiagnosticBase>
+		{
+			new W001IllegalProjectTypeDiagnostic(),
+			// W002 is not a real diagnostic
+			new W010ConfigurationsMismatchDiagnostic(),
+			new W020MicrosoftCSharpDiagnostic(),
+		};
+
+		public void Analyze(Project project, AnalysisOptions options = null)
+		{
+			if (project == null)
+			{
+				throw new ArgumentNullException(nameof(project));
+			}
+
+			options = options ?? new AnalysisOptions();
+
+			if (options.RootDirectory == null)
+			{
+				options.RootDirectory = project.Solution?.FilePath.Directory ?? project.FilePath.Directory;
+			}
+
+			foreach (var diagnostic in Diagnostics)
+			{
+				if (!diagnostic.RunOnAnyProject)
+				{
+					if (diagnostic.RunOnLegacyProject && project.IsModernProject)
+					{
+						continue;
+					}
+
+					if (diagnostic.RunOnModernProject && !project.IsModernProject)
+					{
+						continue;
+					}
+				}
+
+				if (options.SkipDiagnostics.Contains(diagnostic.Id))
+				{
+					continue;
+				}
+
+				diagnostic.Project = project;
```

Comment:
assigning these here means we cannot run analysis in parallel which could be a nice optimization for big solutions. Can we make the analyze call a side effect free function?

Reply:
Fair point. But then it seems too many things might have to be passed to AnalyzeImpl as arguments. Project, Analyzer/Reporter. Well, if the number of these is not expected to grow — why not.

Reply:
I think it has to be
Project (what is analyzed)
Options (how to analyze)
Reporter (how to report)

That seems like a decent understandable set or parameters to me

Reply:
And now that I think of that, it would force the diagnostic implementers to pass Reporter to DiagnosticBase.Report. What about keeping Reporter and Options properties since they are common and immutable during the whole analysis time thus they won't affect parallelization?

Reply:
I don’t think the diagnostic should do the reporting, that’s the job of the analyzer, the diagnostic should just generate a diagnostic and return that to the analyzer.

Reply:
I am familiar on the level of user. :) I haven't experienced the need to create my own diagnostic sets yet.
I believe the diagnostic in the end does as little job as possible. It really just finds violations and calls `Report`. Then the DiagnosticBase actually transforms the call into sth consumable by ConsoleReporter/UnitTestReporter/RazorReporter, the diagnostic impl doesn't have to know how it does that. But the DiagnosticBase doesn't do too much work either. It is still very generic. It calls to analysis property — Reporter that does all heavy lifting.
Analyzer is a pass through indeed. I intended it to be that way so that it only prepares everything and then bluntly calls each diagnostic for all projects. Nothing more, nothing less.
I believe we disagree on the responsibility to report, whether it belongs to Diagnostic or Analyzer. But I still don't understand how transferring the responsibility to Analyzer will simplify matters (from my point of view that would do otherwise).

Reply:
Well but if the reporter moves out of AnalyzerOptions, you’ll have to assign that as well. Maybe I can create a branch based on this one and make a proposal.

Reply:
I’ll see if I can create something tomorrow.

Reply:
I will certainly take all comments into account and force push changes to this branch. Then you can create the proposal (in any repo, I added you as collaborator to my fork if you wish not to clutter this public one) and we will see if it would seem better for all sides of implementers.